File: year2019/day10.py
import numpy as np
from collections import defaultdict
import math
# Cannot use Arctan eg 0,0 and -1,-1 rets 1. 0.0 and 1.1 rets 1. Both are the same
# Use atan2 to compute unique slopes https://en.wikipedia.org/wiki/Atan2
# atan2 takes y and x
class Asteroid:
    def __init__(self,x,y):
        self.x = x
        self.y = y
        self.unique = set()
        self.max = -math.inf
        self.distance = -math.inf

    def __str__(self):
        return "x_coordinate %s, y_coordinate %s" %(self.x,self.y)

    def slope(self,other):
        angle = math.atan2((self.y - other.y),(self.x - other.x))
        self.unique.add(angle)
        return angle

    def set_max(self):
        self.max = len(self.unique)
        return
    def L2_norm(self,other):
        math.sqrt((self.x-other.x)**2+(self.y-other.y)**2)
        return

with open("day10.txt", 'r') as fp:
    galaxy = fp.readlines()

    ast_list = []

    for x,line in enumerate(galaxy):
        for y,ast in enumerate(line):
            if ast == '#':
                ast_list.append(Asteroid(x,y))

station = None
detected = -math.inf

for ast in ast_list:
    for other in ast_list:
        if ast != other:
            angle = ast.slope(other)

    ast.set_max()
    if ast.max > detected:
        detected = ast.max
        station = ast

print(detected) # 299
print(station) # x_coordinate 29, y_coordinate 26

# Our grid is not cartesian plane
#  [0..............N]
#  [.               ]
#  [.               ]
#  [.               ]
#  [.               ]
#  [N               ]

# Fuck it convert to polar coordinates? And set x_coordinate 29, y_coordinate 26 as origin then sort by angle?

File: year2019/day10.jl
function compute_angle(origin,point)

    degree = atan((origin[2] - point[2]),(point[1] - origin[1])) * (180/pi)
    if degree > 90
        degree = 450 - degree
    else
        degree = 90 - degree
    end
    return degree
end

function compute_distance(origin,point)
    distance = sqrt((origin[1] - point[1])^2 + (origin[2] - point[2])^2)
    return distance
end

f = open("day10.txt");
ast_list = []
for (y, line) in enumerate(readlines(f))
    for (x, value) in enumerate(line)
        if value == '#'
            push!(ast_list,[x-1,y-1])
        end

    end
end
detected = -1
station = -1
for ast in ast_list
    detected_angles = Set()
    for other in ast_list
        if ast != other
            push!(detected_angles,compute_angle(ast,other))
        end
    end
    if length(detected_angles) > detected
        global detected
        global station
        detected = length(detected_angles)
        station = ast
    end
end

println(detected)
println(station)

filter!(x->x!=station,ast_list)

angle_dict = Dict()
for ast in ast_list
    angle = compute_angle(station,ast)

    if haskey(angle_dict, angle) == false
        push!(angle_dict,angle=>[ast])
    else
        value = angle_dict[angle]
        append!(value,[ast])
    end
end
iter = 0
for angle in sort([i for i in keys(angle_dict)],by=abs)
    remove = typemax(Float64)
    vaporized = nothing
    for i in angle_dict[angle]
        dist = compute_distance(station,i)
        if dist < remove
            remove = dist
            vaporized = i
        end
    end
    global iter
    iter += 1
    if iter == 200
        solution = vaporized[1]*100 + vaporized[2]
        println("Vaporize $vaporized, solution is $solution")
        break
    end
end
